server/src/service/service.rs

```diff
@@ -0,0 +1,210 @@
+
+use crate::service::collection::Collection;
+use crate::service::session::Session;
+use crate::KEYRING;
+
+#[derive(Serialize, Type, Debug)]
+pub struct Service {
+    pub(self) collections: Vec<Collection>,
+}
+
+#[dbus_interface(name = "org.freedesktop.Secret.Service")]
+impl Service {
+    pub async fn open_session(
+        &self,
+        /*algorithm: &str,*/
+        input: Value<'_>,
+    ) -> fdo::Result<(OwnedValue, OwnedObjectPath)> {
+        // WIP: not complete
+        let session = Session::new(ObjectPath::try_from("/s1").unwrap().into()).await;
+        Ok((
+            OwnedValue::try_from(input).unwrap(), // avoid using unwrap
+            ObjectPath::try_from(session.path().await).unwrap().into(),
+        ))
+    }
+
+    #[dbus_interface(out_args("collection", "prompt"))]
+    pub async fn create_collection(
+        &self,
+        properties: Properties,
+        alias: String,
+        #[zbus(object_server)] object_server: &ObjectServer,
+    ) -> fdo::Result<(ObjectPath, ObjectPath)> {
+        let _attributes = properties.attributes(); // expand and pass these to Collection:new
+        let collection = Collection::new(String::from("temp")).await; // temporarily
+        let path = format!("/org/freedesktop/secrets_/collection/{}", &alias);
+
+        let _ = object_server.at(path.clone(), collection).await;
+        let new_collection_path = ObjectPath::try_from(path).unwrap(); // temporarily
+        let prompt = ObjectPath::default().into(); // temp Prompt
+
+        Ok((new_collection_path, prompt))
+    }
+
+    #[dbus_interface(out_args("unlocked", "locked"))]
+    pub async fn search_items(
+        &self,
+        attributes: HashMap<&str, &str>,
+    ) -> fdo::Result<(Vec<Item>, Vec<Item>)> {
+        let items = match KEYRING.get().unwrap().search_items(&attributes).await {
+            Ok(i) => i,
+            Err(_) => todo!(),
+        };
+
+        let mut unlocked: Vec<Item> = Vec::new();
+        let mut locked: Vec<Item> = Vec::new();
+
+        for item in items {
+            let attributes = item.attributes();
+            if attributes.get("locked").is_some() {
+                // this if condition is probably wrong
+                // how to access &AttributeValue value
+                locked.push(item)
+            } else {
+                unlocked.push(item)
+            }
+        }
+
+        Ok((unlocked, locked))
+    }
+
+    #[dbus_interface(out_args("unlocked", "prompt"))]
+    pub async fn unlock(
+        &mut self,
+        objects: Vec<OwnedObjectPath>,
+    ) -> fdo::Result<(Vec<OwnedObjectPath>, OwnedObjectPath)> {
+        // manage unlock state in memory
+        // when do we need to prompt?
+        let mut unlocked: Vec<OwnedObjectPath> = Vec::new();
+
+        'main: for object in objects {
+            for collection in &mut self.collections {
+                if collection.path().await.as_str() == object.as_str() {
```

Comment:
No need to convert to a string to do a comparison

Reply:
got it, done.